Implemented LBFS/LBFS+

Watercrystal opened the pull request:
As a starting point for the 6-sweep LFBS-based detection algorithm, I implemented LBFS and LBFS+ (this means that LBFS* is still missing, but it should be easy to add to this). I am not a good programmer, so please write a comment/suggestion if something seems bad (and I am pretty sure you are going to find a lot of bad stuff).

pdelvo commented on jgrapht-core/src/main/java/org/jgrapht/alg/intervalgraph/LexBreathFirstSearch.java:
```diff
@@ -0,0 +1,217 @@
+    
+        // return ordering -- thanks Java for not being able to create an array from a list with generic types for some reason
+        V[] orderingArray = null;
+        orderingArray = ordering.toArray(orderingArray);
```
What this does is copying the entire internal array again. Maybe it would be best to not use ArrayList at all and work on a basic array directly, especially because you dont need any feature of the list itself and you know the capacity at the start.

pdelvo commented on jgrapht-core/src/main/java/org/jgrapht/alg/intervalgraph/LexBreathFirstSearch.java:
```diff
@@ -0,0 +1,217 @@
+     * @param startingVertex the starting vertex of the LBFS
+     * @return an array of vertices representing the order in which the vertices were found
+     */
+    public V[] lexBreathFirstSearch(Graph<V, E> graph, V startingVertex)
```
Also no breathing involved in this algorithm as far as I can see 😕

pdelvo commented on jgrapht-core/src/main/java/org/jgrapht/alg/intervalgraph/LexBreathFirstSearch.java:
```diff
@@ -0,0 +1,217 @@
+        int n = graph.vertexSet().size();
+        ArrayList<V> ordering = new ArrayList<V>();
+        
+        String maxLabel = Integer.toString(n);
```
Shouldnt this be updated somewhere in the loop further down? Have you tried the algorithm as is? Does it work?

pdelvo commented on jgrapht-core/src/main/java/org/jgrapht/alg/intervalgraph/LexBreathFirstSearch.java:
```diff
@@ -0,0 +1,217 @@
+     * @param prevOrdering the previous LBFS ordering given as an array
+     * @return an array of vertices representing the order in which the vertices were found
+     */
+    public V[] lexBreathFirstSearch(Graph<V, E> graph, V startingVertex, V[] prevOrdering)
```
This is a lot of copypasta code. Maybe change this to allow prevOrdering to be null which would turn off the sorting (which is the only difference in this algorithm?)

pdelvo replied:
Im not adding any more comments to this algorithm as it is the same as the one above.

pdelvo commented on jgrapht-core/src/main/java/org/jgrapht/alg/intervalgraph/LexBreathFirstSearch.java:
```diff
@@ -0,0 +1,217 @@
+        }
+        
+        // apparantly Java has no indexOf method...
+        private int indexOf(V vertex) throws NoSuchElementException{
```
Nit: Space after the exception

pdelvo commented on jgrapht-core/src/main/java/org/jgrapht/alg/intervalgraph/LexBreathFirstSearch.java:
```diff
@@ -0,0 +1,217 @@
+        private final V[] prevOrdering;
+        
+        public LBFSComparator(V[] prevOrdering) {
+            this.prevOrdering = prevOrdering;
```
Check for null?

pdelvo commented on jgrapht-core/src/main/java/org/jgrapht/alg/intervalgraph/LexBreathFirstSearch.java:
```diff
@@ -0,0 +1,217 @@
+        // apparantly Java has no indexOf method...
+        private int indexOf(V vertex) throws NoSuchElementException{
+            for(int index = 0; index < prevOrdering.length; index++) {
+                if(prevOrdering[index].equals(vertex)) {
```
crashes for nulls in the array

pdelvo commented on jgrapht-core/src/main/java/org/jgrapht/alg/intervalgraph/LexBreathFirstSearch.java:
```diff
@@ -0,0 +1,217 @@
+        }
+        
+        @Override
+        public int compare(V vertex1, V vertex2)
```
This is currently a O(n^2) algorithm which makes sorting O(n^2*n log(n)) = O(n^3 log(n)). Maybe use a map vertex -> position (with a hashmap)

Watercrystal commented:
I have now rewritten the LBFS/LBFS+ stuff to work around the existing implementation rather than my own, which should hopefully make most comments obsolete. Now, the LBFS class just provides an interface for retrieving the ordering from the iterator, which I modified to be able to cope with the additional requirements. The only bit of code that was reused is the comparator (with the requested changes) and the existing class (however, it is just a wrapper now ... but I fixed its name :P).

PhoenixIra commented on jgrapht-core/src/main/java/org/jgrapht/alg/intervalgraph/IOrdering.java:
```diff
@@ -0,0 +1,32 @@
+     * @param sweep the order we want to check if its an I-Order
+     * @param graph the graph we want to check if its an I-Order
+     * @return true, if sweep is an I-Order according to graph
+     */
```
I think, this algorithm has a complexity of O(|sweep|^3) if containsEdge is in O(1). According to the paper, there should also be an algorithm for O(|V|+|E|). We should maybe consider thinking about how to implement a linear time i-order check? (at least after tuesday)

PhoenixIra commented on jgrapht-core/src/main/java/org/jgrapht/traverse/LexBreadthFirstIterator.java:
```diff
@@ -379,4 +419,34 @@ boolean isEmpty() {
+            
+            return (index1 - index2);
+        }
+    }
```
Some basic docs for the comparator would be nice, I guess :D

PhoenixIra commented on jgrapht-core/src/main/java/org/jgrapht/traverse/LexBreadthFirstIterator.java:
```diff
@@ -379,4 +419,34 @@ boolean isEmpty() {
+    
+    class PriorityComparator implements Comparator<V>
+    {
+        private final V[] priority;
```
As @pdelvo mentioned it a commit before, a HashMap should be better here to decrease the compare complexity

PhoenixIra commented on jgrapht-core/src/main/java/org/jgrapht/traverse/LexBreadthFirstIterator.java:
```diff
@@ -74,6 +74,18 @@ public LexBreadthFirstIterator(Graph<V, E> graph) {
+     * @param graph the graph to be iterated.
+     * @param priority the vertex array sorted by their priorities.
+     */
+    public LexBreadthFirstIterator(Graph<V, E> graph, V[] priority) {
```
Do we maybe want to check, if the priority is a valid priority of graph? i.e. check if graph.vertices and priority have the same vertices.

pdelvo commented on jgrapht-core/src/main/java/org/jgrapht/alg/intervalgraph/IntervalGraphRecognizer.java:
```diff
@@ -0,0 +1,142 @@
+public final class IntervalGraphRecognizer {
+
+  /**
+   * check if the given graph is an interval graph
```
It might be good to link the paper this is based on.

pdelvo commented on jgrapht-core/src/main/java/org/jgrapht/alg/intervalgraph/IntervalGraphRecognizer.java:
```diff
@@ -0,0 +1,142 @@
+   * @param graph the graph we want to check if its an I-Order
+   * @return true, if sweep is an I-Order according to graph
+   */
+  private static <V> boolean isIOrdering(V[] sweep, Graph graph) {
```
This implementation has running time Omega(n^3) which is worse than the aim of O(n + m)

pdelvo commented on jgrapht-core/src/main/java/org/jgrapht/alg/intervalgraph/IntervalGraphRecognizer.java:
```diff
@@ -0,0 +1,142 @@
+   * @param <E>   the generic type representing edges
+   * @return
+   */
+  public static <V, E> boolean isIntervalGraph(Graph<V, E> graph) {
```
We have to special case the empty graph (which is an interval graph) and throw an exception if graph is null.

pdelvo commented on jgrapht-core/src/main/java/org/jgrapht/alg/intervalgraph/IntervalGraphRecognizer.java:
```diff
@@ -0,0 +1,142 @@
+   */
+  private static <V> V randomElementOf(Set<V> set) {
+    if (set == null) {
+      throw new IllegalArgumentException("List parameter cannot be null.");
```
Set, not list

pdelvo commented on jgrapht-core/src/main/java/org/jgrapht/alg/intervalgraph/IntervalGraphRecognizer.java:
```diff
@@ -0,0 +1,142 @@
+      throw new IllegalArgumentException("List parameter cannot be null.");
+    }
+
+    int index = new Random().nextInt(set.size());
```
We should store the Random object. Otherwise new Random().nextInt(...), when called a lot in a small loop, might return the same value (as it is initialized with the system time)

pdelvo commented on jgrapht-core/src/main/java/org/jgrapht/alg/intervalgraph/IntervalGraphRecognizer.java:
```diff
@@ -0,0 +1,142 @@
+    return resultArray;
+  }
+
+}
```
Add a new line at the end of the file

pdelvo commented on jgrapht-core/src/main/java/org/jgrapht/alg/intervalgraph/IntervalGraphRecognizer.java:
```diff
@@ -0,0 +1,142 @@
+      throw new IllegalArgumentException("List parameter cannot be null.");
+    }
+
+    V[] resultArray = (V[]) new Object[list.size()];
```
This looks so wrong. Why is Java so broken???

magnificent-tony commented on jgrapht-core/src/main/java/org/jgrapht/alg/intervalgraph/LexBreadthFirstSearch.java:
```diff
@@ -0,0 +1,50 @@
+     * @param startingVertex the starting vertex of the LBFS
+     * @return an array of vertices representing the order in which the vertices were found
+     */
+    public List<V> lexBreadthFirstSearch(Graph<V, E> graph, V startingVertex)
```
startingVertex is not used further in the method.

magnificent-tony commented on jgrapht-core/src/main/java/org/jgrapht/alg/intervalgraph/LexBreadthFirstSearch.java:
```diff
@@ -0,0 +1,50 @@
+     * @return an array of vertices representing the order in which the vertices were found
+     */
+    
+    public List<V> lexBreadthFirstSearch(Graph<V, E> graph, V startingVertex, V[] priority)
```
startingVertex is not used further in the method, but we need it as "in fact" the first vertex to be visited.

Watercrystal commented:
I have now completed most of the functionality, hopefully. I couldnt add support for outputting the interval representation just yet as the interfaces created by the data structure team are not merged with this branch; but all the values needed are computed. I fixed the problem @magnificent-tony found (startingVertex not being used -- good catch!) and improved the comparator (however, I still lack lots of documentation; I tried getting all the stuff we need together first). In addition to that, I implemented @danielmock's suggested I-Ordering algorithm (however I did not check if it actually is correct) which has a linear running time. Moreover, I refactored the code to use hashmaps for basically everything instead of arrays (which dont work very nicely with generics anyways) and standardized the usage of data structures (I hope so at least). 

Some remarks:
- @pdelvo: The choice of random element is not important at all, so even if this were to happen, it is not an issue for us.
- I have not implemented checks for the validity of the orderings yet (but we can do so if you think it is important).

Watercrystal commented:
Refactored the recognizer class such that the algorithm is ran only once upon creating an instance; the results are stored in local fields accessible by getters (as discussed at the last meeting). Further, I added (javadoc) documentation to functions missing it as well as the class itself (complete with a reference to the paper we used).

As requested by @christophgruene, the interval list computed by the algorithm now is sorted by the starting points of the intervals (ascendingly; done in linear time using bucket sort) to allow for a more efficient construction of the interval representation once we merge.

pdelvo commented on jgrapht-core/src/main/java/org/jgrapht/alg/intervalgraph/IntervalGraphRecognizer.java:
```diff
@@ -0,0 +1,273 @@
+    private static <V, E> boolean isIOrdering(HashMap<V, Integer> sweep, Graph<V, E> graph)
+    {
+        // Compute inverse sweep map to quickly find vertices at given indices
+        HashMap<Integer, V> inverseSweep = new HashMap<>();
```
This can be replaced by an array.

Watercrystal replied:
Arrays of generics, unfortunately, are a bit of a pain in the ass in Java, which is why I opted to use HashMaps instead whenever I needed an array.

danielmock replied:
Instead of arrays you can always use an appropriately initialized ArrayList
